db.go

```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
 	cdb := (**C.notmuch_database_t)(unsafe.Pointer(&db.cptr))
 	cerr := C.notmuch_database_open(cpath, cmode, cdb)
 	runtime.SetFinalizer(db, func(db *DB) {
-		C.notmuch_database_destroy(db.toC())
+		C.notmuch_database_destroy(db.cptr)
 	})
 	return db, statusErr(cerr)
 }
 
+// Atomic opens an atomic transaction in the database and calls the callback.
+func (db *DB) Atomic(callback func(*DB)) error {
+	cerr := C.notmuch_database_begin_atomic(db.cptr)
+	if err := statusErr(cerr); err != nil {
+		return err
+	}
+	callback(db)
+	cerr = C.notmuch_database_end_atomic(db.cptr)
+	if err := statusErr(cerr); err != nil {
+		return err
+	}
+	return nil
+}
+
 // NewQuery creates a new query from a string following xapian format.
 func (db *DB) NewQuery(queryString string) *Query {
 	cstr := C.CString(queryString)
 	defer C.free(unsafe.Pointer(cstr))
-	cquery := C.notmuch_query_create(db.toC(), cstr)
+	cquery := C.notmuch_query_create(db.cptr, cstr)
 	query := &Query{
 		cptr: cquery,
 		db:   db,
 	}
 	runtime.SetFinalizer(query, func(q *Query) {
-		C.notmuch_query_destroy(q.toC())
+		C.notmuch_query_destroy(q.cptr)
 	})
 	return query
 }
 
 // Close closes the database.
 func (db *DB) Close() error {
-	cdb := (*C.notmuch_database_t)(db.cptr)
-	cerr := C.notmuch_database_close(cdb)
+	cerr := C.notmuch_database_close(db.cptr)
 	err := statusErr(cerr)
 	return err
 }
 
 // Version returns the database version.
 func (db *DB) Version() int {
-	return int(C.notmuch_database_get_version(db.toC()))
+	return int(C.notmuch_database_get_version(db.cptr))
 }
 
 // LastStatus retrieves last status string for the notmuch database.
 func (db *DB) LastStatus() string {
-	return C.GoString(C.notmuch_database_status_string(db.toC()))
+	return C.GoString(C.notmuch_database_status_string(db.cptr))
 }
 
-func (db *DB) toC() *C.notmuch_database_t {
-	return (*C.notmuch_database_t)(db.cptr)
+// Path return the database path of the database.
+func (db *DB) Path() string {
+	return C.GoString(C.notmuch_database_get_path(db.cptr))
+}
+
+// NeedsUpgrade returns true if the database can be upgraded. This will always
+// return false if the database was opened with DBReadOnly.
+//
+// If this function returns true then the caller may call
+// Upgrade() to upgrade the database.
+func (db *DB) NeedsUpgrade() bool {
+	cbool := C.notmuch_database_needs_upgrade(db.cptr)
+	return int(cbool) != 0
+}
+
+// Upgrade upgrades the current database to the latest supported version. The
+// database must be opened with DBReadWrite.
+func (db *DB) Upgrade() error {
+	return statusErr(C.notmuch_database_upgrade(db.cptr, nil, nil))
+}
+
+// AddMessage adds a new message to the current database or associate an
+// additional filename with an existing message.
+func (db *DB) AddMessage(p string) (*Message, error) {
+	cpath := C.CString(p)
+	defer C.free(unsafe.Pointer(cpath))
+
+	msg := &Message{}
+	cmsg := (**C.notmuch_message_t)(unsafe.Pointer(msg.cptr))
```

Comment:
This looks incorrect to me; I think you need to be taking the address of `msg.cptr`.